_posts/2020-05-29-pixelcnn/pixelcnn_tfprobability.Rmd

```diff
@@ -0,0 +1,326 @@
+
+```{r}
+# bee, bicycle, broccoli, butterfly, cactus,
+# frog, guitar, lightning, penguin, pizza,
+# rollerskates, sea turtle, sheep, snowflake, sun,
+# swan, The Eiffel Tower, tractor, train, tree
+classes <- c(26, 29, 43, 49, 50,
+             125, 134, 172, 218, 225,
+             246, 255, 258, 271, 295,
+             296, 308, 320, 322, 323
+)
+
+classes_tensor <- tf$cast(classes, tf$int64)
+
+train_ds <- train_ds %>%
+  dataset_filter(
+    function(record) tf$reduce_any(tf$equal(classes_tensor, record$label), -1L)
+  )
+
+```
+
+The PixelCNN distribution expects values in the range from 0 to 255 -- no normalization required. Preprocessing then consists
+of just casting pixels and labels each to `float`:
+
+```{r}
+preprocess <- function(record) {
+  record$image <- tf$cast(record$image, tf$float32) 
+  record$label <- tf$cast(record$label, tf$float32)
+  list(tuple(record$image, record$label))
+}
+
+batch_size <- 32
+
+train <- train_ds %>%
+  dataset_map(preprocess) %>%
+  dataset_shuffle(10000) %>%
+  dataset_batch(batch_size)
+```
+
+### Creating the model
+
+We now use [tfd\_pixel\_cnn](https://rstudio.github.io/tfprobability/reference/tfd_pixel_cnn.html) to define what will be the
+loglikelihood used by the model.
+
+```{r}
+dist <- tfd_pixel_cnn(
+  image_shape = c(28, 28, 1),
+  conditional_shape = list(),
+  num_resnet = 5,
+  num_hierarchies = 3,
+  num_filters = 128,
+  num_logistic_mix = 5,
+  dropout_p =.5
+)
+
+image_input <- layer_input(shape = c(28, 28, 1))
+label_input <- layer_input(shape = list())
+log_prob <- dist %>% tfd_log_prob(image_input, conditional_input = label_input)
+```
+
+This custom loglikelihood is added as a loss to the model, and then, the model is compiled with just an optimizer
+specification only. During training, loss first decreased quickly, but improvements from later epochs were smaller.
+
+```{r}
+model <- keras_model(inputs = list(image_input, label_input), outputs = log_prob)
+model$add_loss(-tf$reduce_mean(log_prob))
+model$compile(optimizer = optimizer_adam(lr = .001))
+
+model %>% fit(train, epochs = 10)
+```
+
+To jointly display real and fake images:
+
+```{r}
+for (i in classes) {
+  
+  real_images <- train_ds %>%
+    dataset_filter(
+      function(record) record$label == tf$cast(i, tf$int64)
+    ) %>% 
+    dataset_take(8) %>%
+    dataset_batch(8)
+  it <- as_iterator(real_images)
+  real_images <- iter_next(it)
+  real_images <- real_images$image %>% as.array()
+  real_images <- real_images[ , , , 1]/255
+  
+  generated_images <- dist %>% tfd_sample(8, conditional_input = i)
+  generated_images <- generated_images %>% as.array()
+  generated_images <- generated_images[ , , , 1]/255
+  
+  images <- abind::abind(real_images, generated_images, along = 1)
+  png(paste0("draw_", i, ".png"), width = 8 * 28 * 10, height = 2 * 28 * 10)
+  par(mfrow = c(2, 8), mar = c(0, 0, 0, 0))
+  images %>%
+    purrr::array_tree(1) %>%
+    purrr::map(as.raster) %>%
+    purrr::iwalk(~{plot(.x)})
```

Comment:
Can't we remove `{}`? As in `purrr::iwalk(~plot(.x))`, or even better, `purrr::iwalk(plot)`?

Reply:
that's a lot nicer, thanks!